File: app/internal/lcd/spectrogram.go
package lcd

import (
	"fmt"
	"math"

	"open-sds/app/internal/engine"
)

// Spectrogram ("FFT over time"): a scrolling waterfall of Hann-windowed
// magnitude spectra — X = frequency, Y = time (newest at top), colour = dB.
// One new row + a one-row scroll per frame (docs/spectrogram-plan.md). The LCD
// loop owns it and feeds fresh frames; Render blits its image so Render stays
// pure (screenshot-safe).

// spectrogram plot geometry on the 800×480 panel
const (
	sgX0   = 56     // left of the heatmap (room for dB axis on the left key)
	sgY0   = 22     // top
	sgX1   = W - 18 // right
	sgY1   = H - 26 // bottom (room for the frequency axis)
	sgCols = sgX1 - sgX0
	sgRows = sgY1 - sgY0
	sgFFTN = 2048 // decimate the record to ~this many samples for the row FFT
)

// Spectrogram is the scrolling image plus the last effective Nyquist (for the
// frequency axis). Not safe for concurrent Push; the LCD loop is the only
// writer.
type Spectrogram struct {
	img     *MemSurface
	effNyq  float64 // Hz at the right edge of the heatmap
	rows    int     // rows painted so far (for the "warming up" hint)
	floorDB float64 // colour-map floor (dB below the row peak)
}

// NewSpectrogram allocates the waterfall image.
func NewSpectrogram() *Spectrogram {
	return &Spectrogram{img: NewMemSurface(), floorDB: -60}
}

// heat maps t∈[0,1] to a perceptual black→blue→cyan→green→yellow→red→white
// ramp (RGB565). Monotone in brightness so higher dB always reads "hotter".
func heat(t float64) uint16 {
	// Total in t: a NaN (from a degenerate 0·Inf in the paint math) must not
	// reach int(NaN) → a garbage stops[] index. NaN and the low tail → black.
	if math.IsNaN(t) || t <= 0 {
		return rgb(0, 0, 0)
	}
	if t >= 1 {
		return rgb(255, 255, 255)
	}
	// six segments across the ramp
	const seg = 6.0
	s := t * seg
	i := int(s)
	if i < 0 { // defensive: keep the index in range whatever float slips through
		i = 0
	}
	if i > 5 {
		i = 5
	}
	f := s - float64(i)
	lerp := func(a, b uint8) uint8 { return uint8(float64(a) + (float64(b)-float64(a))*f) }
	// inferno-like ramp: monotone in brightness so higher dB always reads hotter
	stops := [7][3]uint8{
		{0, 0, 0}, {40, 0, 90}, {130, 20, 90}, {200, 50, 20}, {240, 150, 10}, {250, 220, 80}, {255, 255, 255},
	}
	a, b := stops[i], stops[i+1]
	return rgb(lerp(a[0], b[0]), lerp(a[1], b[1]), lerp(a[2], b[2]))
}

// Push computes the frame's magnitude spectrum, scrolls the image down one row,
// and paints the new spectrum as the top row. Ch selects the source channel.
func (sg *Spectrogram) Push(f *engine.Frame, ch int, effNyq float64) {
	if f == nil {
		return
	}
	src := f.C1
	if ch == 1 {
		src = f.C2
	}
	valid := f.Valid
	if valid > len(src) {
		valid = len(src)
	}
	if valid < 32 {
		return
	}
	// Stride the whole record down to ~sgFFTN samples so the spectrum covers the
	// full capture at a USEFUL frequency resolution (like the FFT view) — not the
	// raw Nyquist, which crams a low-frequency signal into the leftmost sliver.
	stride := valid / sgFFTN
	if stride < 1 {
		stride = 1
	}
	mags, peak := spectrumMags(src[:valid], stride)
	if mags == nil || peak <= 0 {
		return
	}
	sg.effNyq = effNyq / float64(stride) // Nyquist shrinks by the decimation

	// scroll the heatmap region down one row (copy each row into the one below)
	rowBytes := W * 2
	for y := sgY1 - 1; y > sgY0; y-- {
		copy(sg.img.Pix[y*rowBytes+sgX0*2:y*rowBytes+sgX1*2],
			sg.img.Pix[(y-1)*rowBytes+sgX0*2:(y-1)*rowBytes+sgX1*2])
	}
	// paint the newest spectrum on the top row: column x → freq bin, dB → colour
	half := len(mags)
	floor := sg.floorDB
	if !(floor < 0) { // NaN or >=0 would make invFloor Inf/NaN → poison the paint
		floor = -60
	}
	invFloor := 1.0 / (-floor)
	for x := 0; x < sgCols; x++ {
		k := x * half / sgCols
		db := 20 * math.Log10(mags[k]/peak+1e-12)
		t := 1 + db*invFloor // db=0 → 1, db=floorDB → 0
		sg.img.SetPixel(sgX0+x, sgY0, heat(t))
	}
	if sg.rows < sgRows {
		sg.rows++
	}
}

// Rows reports how many waterfall rows have been painted (debug/status).
func (sg *Spectrogram) Rows() int { return sg.rows }

// Clear blanks the waterfall.
func (sg *Spectrogram) Clear() {
	for i := range sg.img.Pix {
		sg.img.Pix[i] = 0
	}
	sg.rows = 0
}

// spectrumMags returns the Hann-windowed half-spectrum magnitudes of a real
// record (largest power-of-two ≤ len) and the peak. Mirrors the FFT view and
// the web spectrum().
func spectrumMags(src []uint8, stride int) ([]float64, float64) {
	if stride < 1 {
		stride = 1
	}
	avail := len(src) / stride
	n := 1
	for n*2 <= avail {
		n <<= 1
	}
	if n < 16 {
		return nil, 0
	}
	re := make([]float64, n)
	im := make([]float64, n)
	var mean float64
	for i := 0; i < n; i++ {
		mean += float64(src[i*stride])
	}
	mean /= float64(n)
	for i := 0; i < n; i++ {
		w := 0.5 - 0.5*math.Cos(2*math.Pi*float64(i)/float64(n-1))
		re[i] = (float64(src[i*stride]) - mean) * w
	}
	fftRadix2(re, im)
	half := n / 2
	mags := make([]float64, half)
	peak := 1e-9
	for k := 0; k < half; k++ {
		mags[k] = math.Hypot(re[k], im[k])
		if mags[k] > peak {
			peak = mags[k]
		}
	}
	return mags, peak
}

// drawSpectrogram blits the waterfall image and draws the frequency axis + a dB
// colour key. `sg` is the accumulated image passed via the HUD.
func drawSpectrogram(sf Surface, sg *Spectrogram) {
	if sg == nil || sg.rows == 0 {
		DrawText(sf, W/2-170, H/2, "SPECTROGRAM — FFT over time; needs a triggered signal", colDim, 1)
		return
	}
	// blit the heatmap region
	for y := sgY0; y < sgY1; y++ {
		for x := sgX0; x < sgX1; x++ {
			sf.SetPixel(x, y, sg.img.At(x, y))
		}
	}
	// frequency axis (0 .. effNyq) with a few ticks
	if sg.effNyq > 0 {
		for i := 0; i <= 4; i++ {
			x := sgX0 + i*sgCols/4
			f := sg.effNyq * float64(i) / 4
			DrawText(sf, x-12, sgY1+6, fmtFreq(f), colDim, 1)
			sf.SetPixel(x, sgY1, colAxis)
		}
	}
	// time arrow (newest at top)
	DrawText(sf, 6, sgY0, "new", colDim, 1)
	DrawText(sf, 6, sgY1-10, "old", colDim, 1)
	// dB colour key along the top-right
	for i := 0; i < 40; i++ {
		c := heat(float64(i) / 39)
		fillRect(sf, sgX1-44+i, 6, 1, 8, c)
	}
	DrawText(sf, sgX1-64, 8, "dB", colDim, 1)
	DrawText(sf, sgX0+2, 8, fmt.Sprintf("spgm %.0fdB floor", sg.floorDB), colDim, 1)
}

File: app/internal/web/spectrogram.js
// Spectrogram ("FFT over time") waterfall for the web. Each frame's Hann-
// windowed magnitude spectrum (peaks.js spectrum()) becomes a colour-mapped
// ROW; rows scroll down (newest on top). Colour parity with the Go LCD heat().
// Shared helpers exported for a node parity test.

// sgHeat maps t∈[0,1] to an inferno-like [r,g,b] ramp — monotone in brightness
// so higher dB always reads hotter. Identical stops to the Go heat().
function sgHeat(t) {
  if (t <= 0) return [0, 0, 0];
  if (t >= 1) return [255, 255, 255];
  const stops = [[0, 0, 0], [40, 0, 90], [130, 20, 90], [200, 50, 20], [240, 150, 10], [250, 220, 80], [255, 255, 255]];
  const s = t * 6, i = Math.floor(s), f = s - i;
  const a = stops[i], b = stops[i + 1];
  return [Math.round(a[0] + (b[0] - a[0]) * f), Math.round(a[1] + (b[1] - a[1]) * f), Math.round(a[2] + (b[2] - a[2]) * f)];
}

// sgNew allocates a w×h waterfall (w = frequency columns, h = time rows).
function sgNew(w, h) {
  return { w, h, data: new Uint8ClampedArray(w * h * 4), floorDb: -60, rows: 0, nyq: 0 };
}

// sgPushRow scrolls the image down one row and paints the newest spectrum on
// top. mags/half/peak come from peaks.js spectrum(); nyq is its Nyquist (Hz).
function sgPushRow(sg, mags, half, peak, nyq) {
  const { w, h, data } = sg;
  if (!(half > 0) || !(peak > 0)) return;
  data.copyWithin(w * 4, 0, w * 4 * (h - 1)); // scroll down 1 row
  const invFloor = 1 / (-sg.floorDb);
  for (let x = 0; x < w; x++) {
    const k = Math.min(half - 1, Math.floor(x * half / w));
    const db = 20 * Math.log10(mags[k] / peak + 1e-12);
    const t = Math.max(0, Math.min(1, 1 + db * invFloor));
    const [r, g, b] = sgHeat(t);
    const o = x * 4;
    data[o] = r; data[o + 1] = g; data[o + 2] = b; data[o + 3] = 255;
  }
  if (nyq) sg.nyq = nyq;
  if (sg.rows < h) sg.rows++;
}

function sgClear(sg) { sg.data.fill(0); sg.rows = 0; }

// sgBlit draws the waterfall onto a 2D context sized cw×ch (the ImageData is
// the same size, so it maps 1:1), plus a frequency axis and a dB colour key.
function sgBlit(g, cw, ch, sg, textColor) {
  g.clearRect(0, 0, cw, ch);
  if (!sg || sg.rows === 0) {
    g.fillStyle = textColor || "#9ab"; g.textAlign = "center"; g.font = "11px system-ui";
    g.fillText("FFT over time — needs a triggered signal", cw / 2, ch / 2);
    return;
  }
  // the heatmap occupies the top (ch-16) px; the bottom 16 px carry the axis
  const plotH = ch - 16;
  // draw via a temp canvas so we can scale the ImageData to the plot area
  const tmp = sgBlit._tmp || (sgBlit._tmp = document.createElement("canvas"));
  tmp.width = sg.w; tmp.height = sg.h;
  tmp.getContext("2d").putImageData(new ImageData(sg.data, sg.w, sg.h), 0, 0);
  g.imageSmoothingEnabled = false;
  g.drawImage(tmp, 0, 0, sg.w, sg.h, 0, 0, cw, plotH);
  // frequency axis
  g.fillStyle = textColor || "#9ab"; g.font = "10px system-ui"; g.textAlign = "center"; g.textBaseline = "top";
  for (let i = 0; i <= 4; i++) {
    const x = i * cw / 4;
    const f = sg.nyq * i / 4;
    g.fillText(f >= 1e6 ? (f / 1e6).toPrecision(3).replace(/\.?0+$/, "") + "M" : (f / 1e3).toFixed(0) + "k", Math.min(cw - 14, Math.max(14, x)), plotH + 3);
  }
}

if (typeof module !== "undefined" && module.exports) {
  module.exports = { sgHeat, sgNew, sgPushRow, sgClear };
}
